### PR TITLE
[DNM] msg/async/rdma: support RDMA in systemd

### PR DESCRIPTION
**Pull request by Adirl**
parse ceph.conf to get ms_type
override systemd params as needed if using RDMA messenger 
restore original params on service stop

Signed-off-by: DanielBar-On <danielbo@mellanox.com>

### CODE REVIEW

**Comment by Adirl** [comment]
@liewegas @tchaikov @batrick @yuyuyu101 @DanielBar-On @saritz @smithfarm @mmgaggle 

replacing PR: msg/async/rdma: support RDMA in systemctl #13305
verified on RDMA and TCP cluster

File: systemd/check_rdma.sh [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+rm_config(){
+	for dir in "$@"                                                                            #removing is easier since it will only revert what already exists
+	do
+		sudo sed -i -e 's/PrivateDevices=no/PrivateDevices=yes/g' $dir/override.conf
+		sudo sed -i '/LimitMEMLOCK=infinity/d' $dir/override.conf
+	done
+}
+
+if grep -q "rdma" /etc/ceph/ceph.conf ; then
```
Comment (tchaikov) [inline systemd/check_rdma.sh]: you can use `ceph-conf`command to check a certain option in ceph.conf.

**Comment by Adirl** [comment]
@tchaikov, thanks for the good advice

Done

File: systemd/check_rdma.sh [hunk 2]
```diff
@@ -0,0 +1,46 @@
+#!/bin/bash
+
+sudo systemctl daemon-reload
```
Comment (batrick) [inline systemd/check_rdma.sh]: Why does this need to be done at the start?

File: systemd/check_rdma.sh [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+OSD_DIR="/etc/systemd/system/ceph-osd@.service.d"
+ARG_DIR="/etc/systemd/system/"$INST"@.service.d"
+
+add_config(){
+	for dir in "$@"
+	do
+		sudo mkdir $dir
+		touch $dir/override.conf
+		if ! grep -q "[Service]" $dir/override.conf ; then
+			sudo echo "[Service]"| sudo tee -a $dir/override.conf
```
Comment (batrick) [inline systemd/check_rdma.sh]: sudo in `sudo echo` is unnecessary.

File: systemd/check_rdma.sh [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+if [[ $(ceph-conf ms_type) = *rdma* ]] || [[ $(ceph-conf ms_public_type) = *rdma* ]] || [[ $(ceph-conf ms_cluster_type) = *rdma* ]] ;then
+	if [ "$2" != "clean" ]; then
+		add_config $MON_DIR $ARG_DIR $OSD_DIR
+	else
+		rm_config $MON_DIR $ARG_DIR $OSD_DIR
+	fi
+fi
+
+sudo systemctl daemon-reload
```
Comment (batrick) [inline systemd/check_rdma.sh]: I'd like to see daemon-reload only done if a change is made.

**Comment by smithfarm** [comment]
Do I understand correctly that this is proposing to run an RDMA-specific shell script every time any Ceph systemd unit is started or stopped, regardless of whether the underlying hardware is RDMA or not?

Given that only a small fraction of `total_number_of_Ceph_clusters_in_the_world`, I wonder if there's any other way to achieve the desired end?

File: systemd/check_rdma.sh (hunk 4 above)
Comment (smithfarm) [inline systemd/check_rdma.sh]: Yeah, it looks like this script (which is triggered every time any Ceph-related systemd unit is started *or* stopped) is running daemon-reload twice on every invocation, regardless of whether the script actually does anything or not.

Comment (smithfarm) [inline systemd/check_rdma.sh]: Also, what happens if ceph-conf fails for whatever reason? Will systemd start the daemon?

I'm concerned that this is de facto making all the systemd units dependent on ceph-conf being present and functional?

**Comment by smithfarm** [comment]
Maybe we could have separate `-rdma` versions of all the systemd units? Or could the `ExecStartPre` and `ExecStopPost` actions be conditionalized so the script gets run only if an RDMA system-wide configuration variable is set?

**Comment by liewegas** [comment]
yeah, this does not seem like the right approach.  is it possible to make the behavior dependent on a/etc/{default/sysconfig}/ceph variable?  then it's a one-time config change that ceph-ansible or ceph-deploy or the admin can adjust.

**Comment by Adirl** [comment]
Calculating new route.... thanks

**Comment by yuyuyu101** [comment]
@Adirl do we have new plan on this? some people who try to use rdma will meet this problem

**Comment by Adirl** [comment]
@yuyuyu101 @DanielBar-On 
yes! we have new version ready that give answer to all previous comments.
testing is in progress, so far looks good,
see branch here: https://github.com/Adirl/ceph/commits/systemd 

Adir

**Comment by smithfarm** [comment]
Replaced by #14681 looks like - marking DNM

**Comment by yuyuyu101** [comment]
replaced by https://github.com/ceph/ceph/pull/14681

**Pull request closed without merging**